File: Haskell/app/Models/GameScreen.hs
module Models.GameScreen where

import Data.Foldable (fold)
import Data.List (intercalate)
import Data.Maybe (isJust, isNothing)
import Models.Bird (Bird (Bird))
import qualified Models.Bird as Bird
import Models.GameState (GameState (GameState))
import qualified Models.GameState as GameState
import Models.PipeGroup (PipeGroup (PipeGroup))
import qualified Models.PipeGroup as PipeGroup
import qualified Models.Terminal as Terminal
import Utils.Lists (createMatrix, mapWithIndex)
import Utils.Maybe (justOrDefault)

type ScreenMatrixLine = [String]

type ScreenMatrix = [ScreenMatrixLine]

-- Right now, the render methods depend directly on Bird and PipeGroup. In the
-- future, they will receive a GameState object as parameter.

render :: GameState -> IO ()
render state
  | GameState.screenType state == GameState.PLAYING = renderPlayingScreen state
  | GameState.screenType state == GameState.GAMEOVER = renderGameoverScreen state
  | otherwise = renderPausedScreen state

renderGameoverScreen :: GameState -> IO ()
renderGameoverScreen gameState = do
  terminalWidth <- Terminal.getTerminalWidth
  terminalHeight <- Terminal.getTerminalHeight

  let emptyScreenMatrix = createEmptyScreenMatrix terminalWidth (terminalHeight - 1)
  let populatedScreenMatrix = renderTextToScreenMatrix ["Game Over", "Press <Enter> to play again", "High Score: " ++ (show $ GameState.highestScore gameState)] emptyScreenMatrix

  printScreenMatrix populatedScreenMatrix

renderPausedScreen :: GameState -> IO ()
renderPausedScreen gameState = do
  terminalWidth <- Terminal.getTerminalWidth
  terminalHeight <- Terminal.getTerminalHeight

  let emptyScreenMatrix = createEmptyScreenMatrix terminalWidth (terminalHeight - 1)
  let populatedScreenMatrix = renderTextToScreenMatrix ["Welcome to Flap-PLP", "Press <Enter> to flap your wings", "High Score: " ++ (show $ GameState.highestScore gameState)] emptyScreenMatrix

  printScreenMatrix populatedScreenMatrix

renderPlayingScreen :: GameState -> IO ()
renderPlayingScreen gameState = do
  terminalWidth <- Terminal.getTerminalWidth
  terminalHeight <- Terminal.getTerminalHeight

  let emptyScreenMatrix = createEmptyScreenMatrix terminalWidth (terminalHeight - 1)

  let bird = GameState.bird gameState
  let pipeGroups = GameState.pipeGroups gameState
  let score = GameState.score gameState

  let populatedScreenMatrix =
        renderScoreToScreenMatrix score $
          renderBirdToScreenMatrix bird $
            renderPipeGroupsToScreenMatrix pipeGroups emptyScreenMatrix

  printScreenMatrix populatedScreenMatrix

createEmptyScreenMatrix :: Int -> Int -> ScreenMatrix
createEmptyScreenMatrix width height =
  createMatrix width height (\row column -> " ")

renderBirdToScreenMatrix :: Bird -> ScreenMatrix -> ScreenMatrix
renderBirdToScreenMatrix bird screenMatrix =
  renderObjectToScreenMatrix
    (Bird.originX bird)
    (Bird.originY bird)
    (Bird.toString bird)
    screenMatrix

renderPipeGroupsToScreenMatrix :: [PipeGroup] -> ScreenMatrix -> ScreenMatrix
renderPipeGroupsToScreenMatrix pipeGroups screenMatrix
  | null pipeGroups = screenMatrix
  | otherwise =
    renderPipeGroupsToScreenMatrix
      (tail pipeGroups)
      (renderPipeGroupToScreenMatrix (head pipeGroups) screenMatrix)

renderPipeGroupToScreenMatrix :: PipeGroup -> ScreenMatrix -> ScreenMatrix
renderPipeGroupToScreenMatrix pipeGroup screenMatrix =
  renderObjectToScreenMatrix
    (PipeGroup.originX pipeGroup)
    (PipeGroup.originY pipeGroup)
    (PipeGroup.toString pipeGroup)
    screenMatrix

renderScoreToScreenMatrix :: Int -> ScreenMatrix -> ScreenMatrix
renderScoreToScreenMatrix score screenMtx = renderObjectToScreenMatrix scoreOriginX 0 scoreText screenMtx
  where
    scoreText = "Score: " ++ (show score)
    scoreTextLength = length scoreText
    scoreOriginX = ((getWidthScreenMatrix screenMtx) - scoreTextLength) `div` 2

renderObjectToScreenMatrix :: Int -> Int -> String -> ScreenMatrix -> ScreenMatrix
renderObjectToScreenMatrix originX originY objectString screenMatrix
  | originX >= getWidthScreenMatrix screenMatrix = screenMatrix
  | originX + objectWidth <= 0 = screenMatrix
  | originY >= getHeightScreenMatrix screenMatrix = screenMatrix
  | originY + objectHeight <= 0 = screenMatrix
  | otherwise = mapWithIndex
    ( \screenMatrixLine lineY ->
        renderObjectToScreenMatrixLine
          originX
          originY
          objectStringLines
          objectHeight
          screenMatrixLine
          lineY
    )
    screenMatrix
  where
    objectStringLines = lines objectString
    objectHeight = length objectStringLines
    objectWidth = maximum (map length objectStringLines)

renderObjectToScreenMatrixLine :: Int -> Int -> [String] -> Int -> ScreenMatrixLine -> Int -> ScreenMatrixLine
renderObjectToScreenMatrixLine originX originY objectStringLines objectHeight screenMatrixLine lineY
  | lineY < originY || lineY > originY + objectHeight = screenMatrixLine
  | otherwise =
    mapWithIndex
      ( \cellElement cellX ->
          renderObjectToScreenMatrixLineCell
            originX
            originY
            objectStringLines
            objectHeight
            cellX
            lineY
            cellElement
      )
      screenMatrixLine

renderObjectToScreenMatrixLineCell :: Int -> Int -> [String] -> Int -> Int -> Int -> String -> String
renderObjectToScreenMatrixLineCell originX originY objectStringLines objectHeight cellX lineY cellElement =
  characterForCell
  where
    characterForCell = justOrDefault objectCharacterForCell cellElement
    objectCharacterForCell =
      if isJust maybeObjectMatchingLine
        && cellX >= originX
        && cellX < (originX + length objectMatchingLine)
        then Just [objectMatchingLine !! (cellX - originX)]
        else Nothing
    objectMatchingLine = justOrDefault maybeObjectMatchingLine []
    maybeObjectMatchingLine =
      if lineY >= originY
        && lineY < (originY + objectHeight)
        then Just (objectStringLines !! (lineY - originY))
        else Nothing

printScreenMatrix :: ScreenMatrix -> IO ()
printScreenMatrix screenMatrix =
  putStr (formatScreenMatrixToRender screenMatrix)

formatScreenMatrixToRender :: ScreenMatrix -> String
formatScreenMatrixToRender screenMatrix =
  intercalate "\n" (map formatScreenMatrixLineToRender screenMatrix)

formatScreenMatrixLineToRender :: ScreenMatrixLine -> String
formatScreenMatrixLineToRender = intercalate ""

renderTextToScreenMatrix :: [String] -> ScreenMatrix -> ScreenMatrix
renderTextToScreenMatrix (string : []) screenMtx = renderObjectToScreenMatrix stringStartX stringStartY string screenMtx
  where
    stringLength = length string
    stringStartX = (getWidthScreenMatrix screenMtx - stringLength) `div` 2
    stringStartY = (getHeightScreenMatrix screenMtx) `div` 2
renderTextToScreenMatrix (h : t) screenMtx = renderTextToScreenMatrix t (renderTextToScreenMatrixRecursive [h] ((length t) * 2) screenMtx)

renderTextToScreenMatrixRecursive :: [String] -> Int -> ScreenMatrix -> ScreenMatrix
renderTextToScreenMatrixRecursive (string : []) spacing screenMtx = renderObjectToScreenMatrix stringStartX stringStartY string screenMtx
  where
    stringLength = length string
    stringStartX = (getWidthScreenMatrix screenMtx - stringLength) `div` 2
    stringStartY = ((getHeightScreenMatrix screenMtx) `div` 2) - spacing

getHeightScreenMatrix :: ScreenMatrix -> Int
getHeightScreenMatrix screenMtx = length screenMtx

getWidthScreenMatrix :: ScreenMatrix -> Int
getWidthScreenMatrix screenMtx = length $ head screenMtx

File: Prolog/models/gameScreen.pl
:- module(gameScreen, [render/1]).

:- use_module(terminal).
:- use_module(bird).
:- use_module(gameState).
:- use_module('../utils/list').
:- use_module('../utils/strings').

render(GameState):-
  gameState:screenType(GameState,ScreenType),
  renderScreen(ScreenType,GameState).

renderScreen('paused-screen',GameState):-
  terminal:getSize(TerminalWidth, TerminalHeight),

  ScreenMatrixWidth is TerminalWidth,
  ScreenMatrixHeight is TerminalHeight - 1,
  createEmptyScreenMatrix(ScreenMatrixWidth, ScreenMatrixHeight, EmptyScreenMatrix),

  gameState:highestScore(GameState,HighestScore),
  string_concat("High Score: ", HighestScore, LastTextLine),
  
  renderTextToScreenMatrix(["Welcome to Flap-PLP", "Press <Enter> to flap your wings", LastTextLine],EmptyScreenMatrix,ScreenMatrixWithText),
  
  printScreenMatrix(ScreenMatrixWithText).

renderScreen('game-over-screen',GameState):-
  terminal:getSize(TerminalWidth, TerminalHeight),

  ScreenMatrixWidth is TerminalWidth,
  ScreenMatrixHeight is TerminalHeight - 1,
  createEmptyScreenMatrix(ScreenMatrixWidth, ScreenMatrixHeight, EmptyScreenMatrix),

  gameState:highestScore(GameState,HighestScore),
  string_concat("High Score: ", HighestScore, LastTextLine),
  
  renderTextToScreenMatrix(["Game Over", "Press <Enter> to play again", LastTextLine],EmptyScreenMatrix,ScreenMatrixWithText),
  
  printScreenMatrix(ScreenMatrixWithText).

renderScreen('playing-screen',GameState):-
  terminal:getSize(TerminalWidth, TerminalHeight),

  ScreenMatrixWidth is TerminalWidth,
  ScreenMatrixHeight is TerminalHeight - 1,
  createEmptyScreenMatrix(ScreenMatrixWidth, ScreenMatrixHeight, EmptyScreenMatrix),

  gameState:bird(GameState,Bird),
  renderBirdToScreenMatrix(Bird, EmptyScreenMatrix, ScreenMatrixWithBird),
  
  gameState:score(GameState,Score),
  renderScoreToScreenMatrix(Score,ScreenMatrixWithBird,ScreenMatrixWithScore),
  printScreenMatrix(ScreenMatrixWithScore).

renderScoreToScreenMatrix(Score,ScreenMatrix,ScreenMatrixWithScore):-
  string_concat("Score: ",Score,ScoreText),
  atom_length(ScoreText, TextLength),
  widthScreenMatrix(ScreenMatrix,Width),
  OriginX is (Width - TextLength) // 2,
  renderObject(OriginX, 0, ScoreText, ScreenMatrix, ScreenMatrixWithScore).

createEmptyScreenMatrix(ScreenMatrixWidth, ScreenMatrixHeight, EmptyScreenMatrix):-
  list:createMatrix(ScreenMatrixWidth, ScreenMatrixHeight, " ", EmptyScreenMatrix).

renderBirdToScreenMatrix(Bird, ScreenMatrix, ScreenMatrixWithBird):-
  bird:originX(Bird, OriginX),
  bird:originY(Bird, OriginY),
  bird:toString(Bird, BirdAsString),
  renderObject(OriginX, OriginY, BirdAsString, ScreenMatrix, ScreenMatrixWithBird).

renderTextToScreenMatrix([Text|[]],EmptyScreenMatrix,ScreenMatrixWithText):-
  atom_length(Text, TextLength),
  widthScreenMatrix(EmptyScreenMatrix,Width),
  heightScreenMatrix(EmptyScreenMatrix,Height),
  OriginX is (Width - TextLength) // 2,
  OriginY is Height // 2,
  renderObject(OriginX, OriginY, Text,EmptyScreenMatrix,ScreenMatrixWithText).

renderTextToScreenMatrix([TextLine|RemainingTextLines],EmptyScreenMatrix,ScreenMatrixWithText):-
  length(RemainingTextLines,RemainingLinesLength),
  Spacing is RemainingLinesLength * 2,
  renderTextToScreenMatrixRecursive(TextLine,Spacing,EmptyScreenMatrix,ScreenMatrixWithRecursiveText),
  renderTextToScreenMatrix(RemainingTextLines,ScreenMatrixWithRecursiveText,ScreenMatrixWithText).

renderTextToScreenMatrixRecursive(TextLine,Spacing,EmptyScreenMatrix,ScreenMatrixWithText):-
  atom_length(TextLine, TextLength),
  widthScreenMatrix(EmptyScreenMatrix,Width),
  heightScreenMatrix(EmptyScreenMatrix,Height),
  OriginX is (Width - TextLength) // 2,
  OriginY is (Height // 2) - Spacing,
  renderObject(OriginX,OriginY,TextLine,EmptyScreenMatrix,ScreenMatrixWithText).

renderObject(OriginX, OriginY, ObjectString, ScreenMatrix, RenderedScreenMatrix):-
  widthScreenMatrix(ScreenMatrix, WidthScreenMatrix),
  heightScreenMatrix(ScreenMatrix, HeightScreenMatrix),

  strings:lines(ObjectString, ObjectStringLines),
  objectWidth(ObjectStringLines, ObjectWidth),
  objectHeight(ObjectStringLines, ObjectHeight),

  (
    (
      OriginX >= WidthScreenMatrix;
      (OriginX + ObjectWidth) =< 0;
      OriginY >= HeightScreenMatrix;
      (OriginY + ObjectHeight) =< 0
    ) -> RenderedScreenMatrix = ScreenMatrix;

    renderObjectLines(OriginX, OriginY, ObjectStringLines, ObjectHeight, 0, ScreenMatrix, RenderedScreenMatrix)
  ).

renderObjectLines(_OriginX, _OriginY, _ObjectStringLines, _ObjectHeight, _LineY, [], []).
renderObjectLines(
  OriginX,
  OriginY,
  ObjectStringLines,
  ObjectHeight,
  LineY,
  ScreenMatrix,
  RenderedScreenMatrix
):-
  [ScreenMatrixLine | TailScreenMatrixLines] = ScreenMatrix,

  renderObjectLine(
    OriginX,
    OriginY,
    ObjectStringLines,
    ObjectHeight,
    ScreenMatrixLine,
    0,
    LineY,
    RenderedScreenMatrixLine
  ),

  NextLineY is LineY + 1,
  renderObjectLines(
    OriginX,
    OriginY,
    ObjectStringLines,
    ObjectHeight,
    NextLineY,
    TailScreenMatrixLines,
    TailRenderedScreenMatrixLines
  ),

  RenderedScreenMatrix = [RenderedScreenMatrixLine | TailRenderedScreenMatrixLines].

renderObjectLine(_OriginX, _OriginY, _ObjectStringLines, _ObjectHeight, [], _CellX, _LineY, []):- !.
renderObjectLine(
  OriginX,
  OriginY,
  ObjectStringLines,
  ObjectHeight,
  ScreenMatrixLine,
  CellX,
  LineY,
  RenderedScreenMatrixLine
):-
  LineY >= OriginY,
  LineY < OriginY + ObjectHeight,

  [CellElement | TailCellElements] = ScreenMatrixLine,

  renderObjectLineCell(
    OriginX,
    OriginY,
    ObjectStringLines,
    ObjectHeight,
    CellX,
    LineY,
    CellElement,
    RenderedCellElement
  ),

  NextCellX is CellX + 1,
  renderObjectLine(
    OriginX,
    OriginY,
    ObjectStringLines,
    ObjectHeight,
    TailCellElements,
    NextCellX,
    LineY,
    TailRenderedCellElements
  ),

  RenderedScreenMatrixLine = [RenderedCellElement | TailRenderedCellElements],
  !.
renderObjectLine(
  _OriginX,
  _OriginY,
  _ObjectStringLines,
  _ObjectHeight,
  ScreenMatrixLine,
  _CellX,
  _LineY,
  ScreenMatrixLine
).

renderObjectLineCell(
  OriginX,
  OriginY,
  ObjectStringLines,
  _ObjectHeight,
  CellX,
  LineY,
  _CellElement,
  RenderedCellElement
):-
  CellX >= OriginX,

  ObjectMatchingLineIndex is LineY - OriginY,

  nth0(ObjectMatchingLineIndex, ObjectStringLines, ObjectMatchingLine),
  strings:size(ObjectMatchingLine, ObjectMatchingLineLength),

  ObjectCharacterForCellIndex is CellX - OriginX,
  CellX < OriginX + ObjectMatchingLineLength,

  strings:chartAt(ObjectMatchingLine, ObjectCharacterForCellIndex, RenderedCellElement),
  !.
renderObjectLineCell(
  _OriginX,
  _OriginY,
  _ObjectStringLines,
  _ObjectHeight,
  _CellX,
  _LineY,
  CellElement,
  CellElement
).

objectWidth(ObjectStringLines, ObjectWidth):-
  maxObjectStringLineWidth(ObjectStringLines, ObjectWidth).

objectHeight(ObjectStringLines, ObjectWidth):-
  length(ObjectStringLines, ObjectWidth).

maxObjectStringLineWidth([], 0).
maxObjectStringLineWidth([Line | TailLines], MaxWidth):-
  maxObjectStringLineWidth(TailLines, MaxTailWidth),
  strings:size(Line, LineWidth),
  MaxWidth is max(LineWidth, MaxTailWidth).

printScreenMatrix(ScreenMatrix):-
  formatScreenMatrixToRender(ScreenMatrix, FormattedScreenMatrix),
  write(FormattedScreenMatrix),
  ttyflush.

formatScreenMatrixToRender(ScreenMatrix, FormattedScreenMatrix):-
  formatScreenMatrixRows(ScreenMatrix, FormattedRows),
  strings:join(FormattedRows, '\n', FormattedScreenMatrix).

formatScreenMatrixRows([], []).
formatScreenMatrixRows([Row | TailRows], [FormattedRow | FormattedTailRows]):-
  formatScreenMatrixRows(TailRows, FormattedTailRows),
  strings:join(Row, '', FormattedRow).

heightScreenMatrix(ScreenMatrix, Height):- length(ScreenMatrix, Height).

widthScreenMatrix([], 0).
widthScreenMatrix([FirstRow | _], Width):- length(FirstRow, Width).
